galsim/sed.py

```diff
@@ -47,78 +68,113 @@ class SED(object):
+    consequence of using `fast=True` is that __call__ can not accept `astropy.units.Quantity`s in
+    this case.
 
     @param spec          Function defining the z=0 spectrum at each wavelength.  See above for
                          valid options for this parameter.
     @param wave_type     String specifying units for wavelength input to `spec`.
     @param flux_type     String specifying what type of spectral density `spec` represents.  See
                          above for valid options for this parameter.
     @param redshift      Optionally shift the spectrum to the given redshift. [default: 0]
+    @param fast          Convert units on initialization instead of on __call__. [default: True]
     """
-    def __init__(self, spec, wave_type, flux_type, redshift=0.,
-                 _blue_limit=None, _red_limit=None, _wave_list=None, _spec=None ):
+    def __init__(self, spec, wave_type, flux_type, redshift=0., fast=True,
+                 _blue_limit=None, _red_limit=None, _wave_list=None,
+                 _spectral=None, _dimensionless=None):
 
         self._orig_spec = spec  # Save this for pickling
-        self._spec = _spec      # This is orig_spec turned into a function
-
+        # If these are known at construction time, use them.
+        self._spectral = _spectral
+        self._dimensionless = _dimensionless
+
+        if isinstance(wave_type, str):
+            if wave_type.lower() in ['nm', 'nanometer', 'nanometers']:
+                wave_type = units.nm
+            elif wave_type.lower() in ['a', 'ang', 'angstrom', 'angstroms']:
+                wave_type = units.AA
+            else:
+                raise ValueError("Unknown wave_type '{0}'".format(wave_type))
         self.wave_type = wave_type
-        self.flux_type = flux_type
-        self.redshift = redshift
 
-        # Figure out wavelength type
-        if self.wave_type.lower() in ['nm', 'nanometer', 'nanometers']:
-            self.wave_factor = 1.0
-        elif self.wave_type.lower() in ['a', 'ang', 'angstrom', 'angstroms']:
-            self.wave_factor = 10.0
-        else:
-            raise ValueError("Unknown wave_type '{0}'".format(self.wave_type))
+        if isinstance(flux_type, str):
+            if flux_type.lower() == 'flambda':
+                flux_type = units.erg / (units.s * self.wave_type * units.cm**2)
+            elif flux_type.lower() == 'fphotons':
+                flux_type = units.astrophys.photon/(units.s * units.cm**2 * self.wave_type)
+            elif flux_type.lower() == 'fnu':
+                flux_type = units.erg / (units.s * units.Hz * units.cm**2)
+            elif flux_type.lower() == '1':
+                flux_type = units.dimensionless_unscaled
+            else:
+                raise ValueError("Unknown flux_type '{0}'".format(flux_type))
+        self.flux_type = flux_type
+        if not (self.dimensionless or self.spectral):
```

Comment:
Nice simplification.  I changed it a little bit to avoid calling the check_* functions if possible.  Turns out that the `is_equivalent(..., spectral_density())` significantly slows down the code (at least the unit tests) if you run it for every new SED.  That's why there's a `_spectral` init kwarg if you know in advance that you're creating a spectral SED.